src/lib/common/LineMeasureController.svelte

```diff
@@ -0,0 +1,167 @@
+<script lang="ts">
+  import { lineString, type Feature, type LineString } from "@turf/helpers";
+  import length from "@turf/length";
+  import { colors } from "colors";
+  import Pin from "lib/critical_entry/Pin.svelte";
+  import { SecondaryButton } from "lib/govuk";
+  import { LngLat, MapMouseEvent } from "maplibre-gl";
+  import { map } from "stores";
+  import { CollapsibleCard } from ".";
+
+  let isInactive = true;
+  let waypoints: any[] = [];
+  let lineToMeasure: Feature<LineString> | undefined = undefined;
+  let isShiftDown = false;
+  let nextId = 0;
+  const layerName = "measurement-line";
+
+  function handleMapClickEvent(e: MapMouseEvent) {
+    if (isInactive) {
+      return;
+    }
+
+    if (!isShiftDown) {
+      waypoints.push({
+        id: nextId++,
+        lngLat: e.lngLat,
+      });
+      waypoints = waypoints;
+      waypointsUpdated();
+    }
+    if (isShiftDown && waypoints.length > 0) {
+      findAndRemoveNeareastWaypoint(e.lngLat);
+
+      waypointsUpdated();
+    }
+  }
+
+  function findAndRemoveNeareastWaypoint(clickLocation: LngLat) {
+    const maxLngLatDistance = new LngLat(0, -90).distanceTo(new LngLat(0, 90));
+    let currentShortestDistance = maxLngLatDistance;
+    let waypointToRemove: LngLat | undefined = undefined;
+
+    waypoints.forEach((waypoint) => {
+      const waypointLngLat = waypoint.lngLat;
+      if (currentShortestDistance >= clickLocation.distanceTo(waypointLngLat)) {
+        waypointToRemove = waypointLngLat;
+        currentShortestDistance = clickLocation.distanceTo(waypointLngLat);
+      }
+    });
+
+    if (waypointToRemove) {
+      waypoints = waypoints.filter((waypoint) => {
+        const waypointLngLat = waypoint.lngLat;
+        return (
+          waypointLngLat.lng != waypointToRemove?.lng ||
+          waypointLngLat.lat != waypointToRemove?.lat
+        );
+      });
+    }
+  }
+
+  function enableMeasurement() {
+    isInactive = false;
+    $map.on("click", handleMapClickEvent);
+  }
+
+  function disableMeasurement() {
+    isInactive = true;
+    $map.off("click", handleMapClickEvent);
+    waypoints = [];
+    removeLineFromMap();
+  }
+
+  function keyDown(keyDownEvent: KeyboardEvent) {
+    if (keyDownEvent.key === "Shift") isShiftDown = true;
+  }
+
+  function keyUp(keyUpEvent: KeyboardEvent) {
+    if (keyUpEvent.key === "Shift") isShiftDown = false;
+  }
+
+  function waypointsUpdated() {
+    if (waypoints.length < 2) {
+      removeLineFromMap();
+      return;
+    }
+
+    const waypointsAsPosition = waypoints.map((waypoint) => {
+      return [waypoint.lngLat.lng, waypoint.lngLat.lat];
+    });
+    lineToMeasure = lineString(waypointsAsPosition, {});
+    if (lineToMeasure.properties)
+      lineToMeasure.properties.length = length(lineToMeasure) * 1000;
+    updateLinestring();
+  }
+
+  function removeLineFromMap() {
+    if ($map.getSource(layerName)) {
```

Comment:
Constantly adding / removing the source and layer is very confusing to keep track of... the style we use in the rest of the codebase is to declare the source/layer once using the `overwrite` methods, and then `setData`